### PR TITLE
enhance mul_op input error message test=develop

### PR DESCRIPTION
**Pull request by lijianshe02**
fluid.layers.mul:

- Add input type check, error info as follows:

```
import paddle.fluid as fluid
from paddle.fluid import Program, program_guard
with program_guard(Program(), Program()):
    x = np.random.random((2, 4)).astype("float32")
    y = fluid.data(name='x', shape=[3,4],dtype='float32')
    fluid.layers.mul(x=x, y=y)
```
>TypeError: The type of 'x' in mul must be Variable, but received <type 'numpy.ndarray'>

- Add input type check, error info as follows:

```
import paddle.fluid as fluid
from paddle.fluid import Program, program_guard
with program_guard(Program(), Program()):
    x = fluid.data(name='x', shape=[2,4],dtype='float32')
    y = np.random.random((3, 4)).astype("float32")
    fluid.layers.mul(x=x, y=y)
```
>TypeError: The type of 'y' in mul must be Variable, but received <type 'numpy.ndarray'>

- Add data type check, error info as follows:
```
import paddle.fluid as fluid
from paddle.fluid import Program, program_guard
with program_guard(Program(), Program()):
    x = fluid.data(name='x', shape=[3,4],dtype='int32')
    y = fluid.data(name='y', shape=[3,4],dtype='float32')
    fluid.layers.mul(x=x, y=y)
```
>TypeError: The data type of 'x' in mul must be float32 or float64, but received int32

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -13543,6 +13543,23 @@ def mul(x, y, x_num_col_dims=1, y_num_col_dims=1, name=None):
...
+        raise TypeError(
+            "The type of 'x' in mul must be Variable, but received %s" %
+            (type(x)))
+    if not isinstance(y, Variable):
+        raise TypeError(
+            "The type of 'y' in mul must be Variable, but received %s" %
+            (type(y)))
+    if convert_dtype(x.dtype) not in ['float32', 'float64']:
+        raise TypeError(
+            "The data type of 'x' in mul must be float16, float32 or float64, but received %s."
```
Comment (luotao1) [inline python/paddle/fluid/layers/nn.py]: 13556行，这里X可以是float16么？

Comment (lijianshe02) [reply python/paddle/fluid/layers/nn.py]: 改了，不可以

**Review by luotao1 — APPROVED** [review]
LGTM

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -13543,6 +13543,23 @@ def mul(x, y, x_num_col_dims=1, y_num_col_dims=1, name=None):
...
+        raise TypeError(
+            "The type of 'x' in mul must be Variable, but received %s" %
+            (type(x)))
+    if not isinstance(y, Variable):
+        raise TypeError(
+            "The type of 'y' in mul must be Variable, but received %s" %
+            (type(y)))
+    if convert_dtype(x.dtype) not in ['float32', 'float64']:
+        raise TypeError(
+            "The data type of 'x' in mul must be float32 or float64, but received %s."
```
Comment (luotao1) [inline python/paddle/fluid/layers/nn.py:13556]: mul_op是支持float16的，请按#20346 修改，并加上单测。
```
REGISTER_OP_CUDA_KERNEL(
    mul_grad, ops::MulGradKernel<plat::CUDADeviceContext, float>,
    ops::MulGradKernel<plat::CUDADeviceContext, double>,
    ops::MulGradKernel<plat::CUDADeviceContext, plat::float16>);
```

**Pull request merged**